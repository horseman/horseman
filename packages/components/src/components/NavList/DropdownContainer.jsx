import styled from "styled-components";

// Will absolutely position the container undernath the nav list.
const DropdownContainer = styled.div`
  position: absolute;
  top: 100%;
  left: 0;
  z-index: 10;
`;

export default DropdownContainer;
