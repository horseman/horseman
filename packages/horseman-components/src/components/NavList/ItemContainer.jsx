import styled from "styled-components";

/**
 * The ItemContainer is responsible for spacing out the NavItem components
 * within the NavList
 */
const ItemContainer = styled.li`
  position: relative;
  overflow: visible;
  margin-right: 1em;

  &:last-child {
    margin-right: 0;
  }
`;

export default ItemContainer;
