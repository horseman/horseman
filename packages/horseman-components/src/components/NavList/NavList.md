The navList is a grouping of [NavItems](#navitem).

### Examples
```
const data = [{
  to: "#",
  text: "Tours",
  opened: true,
  subnav: [
    {
      to: "#",
      text: "Submenu Item",
    },
    {
      to: "#",
      text: "Another Submenu Item",
    },
  ],
},{
  to: "#",
  text: "Park Guide",
},{
  hasChildren: true,
  to: "#",
  text: "About",
},{
  to: "#",
  text: "Contact",
},];

<section style={{backgroundColor:"#3c83b8", display:"inline-block"}}>
  <NavList navItems={data} />
</section>
```
