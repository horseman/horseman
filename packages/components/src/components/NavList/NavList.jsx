import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import onClickOutside from "react-onclickoutside";

import ItemContainer from "./ItemContainer";
import DropdownContainer from "./DropdownContainer";

// This is the outer unordered list container. It displays flex and contains
// all the accompanying NavItem components on a single line.
const StyledNavList = styled.ul`
  align-items: center;
  border: 0;
  display: flex;
  font: inherit;
  list-style: none;
  margin: 0;
  padding: 0;
`;

const NavList = ({ navItem, dropdown, nav, toggleDropdown, ...rest }) => {
  // Iterate through each navItem and create out list item structure.
  const NavComponent = navItem;
  const DropdownComponent = dropdown;
  const items = nav.map(item => {
    const hasChildren = item.subnav && item.subnav.length > 0;

    // We want the dropdown container to toggle the dropdown when we click
    // outside of the item itself. This HOC will handle adding the appropriate
    // events to the document.
    const EnhancedDropdownContainer = onClickOutside(DropdownContainer, {
      handleClickOutside: () => e => toggleDropdown(e, item),
    });

    return (
      <ItemContainer key={item.text}>
        <NavComponent
          {...item}
          onClick={e =>
            hasChildren && !item.open ? toggleDropdown(e, item) : () => {}
          }
        />
        {item.open && (
          <EnhancedDropdownContainer eventTypes={["click"]}>
            <DropdownComponent nav={item.subnav} />
          </EnhancedDropdownContainer>
        )}
      </ItemContainer>
    );
  });

  return <StyledNavList {...rest}>{items}</StyledNavList>;
};

NavList.propTypes = {
  /**
   * Array of Navigation Items
   */
  nav: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * Drive the dropdown state
       */
      subnav: PropTypes.arrayOf(
        PropTypes.shape({
          to: PropTypes.string.isRequired,
          text: PropTypes.string.isRequired,
        }),
      ),

      /**
       * Navigation Item Anchor URL, If given this nav item will be a Link to
       * the desired location.
       */
      to: PropTypes.string,

      /**
       * Navigation Item Text
       */
      text: PropTypes.string.isRequired,
    }),
  ).isRequired,

  /**
   * Will be rendered in the slot for the item
   */
  navItem: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.node,
    PropTypes.element,
  ]).isRequired,

  /**
   * Will be rendered when a dropdown should be displayed.
   */
  dropdown: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.node,
    PropTypes.element,
  ]).isRequired,

  /**
   * This function will be called when a NavItem, having a subnav, is clicked
   */
  toggleDropdown: PropTypes.func,
};

export default NavList;
