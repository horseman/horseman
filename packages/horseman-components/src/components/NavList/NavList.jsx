import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import onClickOutside from "react-onclickoutside";

import ItemContainer from "./ItemContainer";
import DropdownContainer from "./DropdownContainer";

// This is the outer unordered list container. It displays flex and contains
// all the accompanying NavItem components on a single line.
const StyledNavList = styled.ul`
  align-items: center;
  border: 0;
  display: flex;
  font: inherit;
  list-style: none;
  margin: 0;
  padding: 0;
`;

const NavList = ({ navItem, dropdown, nav, toggleDropdown, ...rest }) => {
  // Iterate through each navItem and create out list item structure.
  const items = nav.map(item => {
    const hasChildren = item.subnav && item.subnav.length > 0;

    // We want the dropdown container to toggle the dropdown when we click
    // outside of the item itself. This HOC will handle adding the appropriate
    // events to the document.
    const EnhancedDropdownContainer = onClickOutside(DropdownContainer, {
      handleClickOutside: () => e => toggleDropdown(e, item),
    });

    return (
      <ItemContainer key={item.text}>
        <navItem
          {...item}
          onClick={e =>
            hasChildren && !item.open ? toggleDropdown(e, item) : () => {}
          }
        />
        {item.open && (
          <EnhancedDropdownContainer eventTypes={["click"]}>
            <dropdown nav={item.subnav} />
          </EnhancedDropdownContainer>
        )}
      </ItemContainer>
    );
  });

  return <StyledNavList {...rest}>{items}</StyledNavList>;
};

NavList.propTypes = {
  /**
   * Array of Navigation Items
   */
  nav: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * Drive the dropdown state
       */
      subnav: PropTypes.arrayOf(
        PropTypes.shape({
          to: PropTypes.string.isRequired,
          text: PropTypes.string.isRequired,
        }),
      ),

      /**
       * Navigation Item Anchor URL, If given this nav item will be a Link to
       * the desired location.
       */
      to: PropTypes.string,

      /**
       * Navigation Item Text
       */
      text: PropTypes.string.isRequired,
    }),
  ).isRequired,

  /**
   * This function will be called when a NavItem, having a subnav, is clicked
   */
  toggleDropdown: PropTypes.func,
};

export default NavList;
