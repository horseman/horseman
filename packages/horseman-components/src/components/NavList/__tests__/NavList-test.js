import React from "react";
import { shallow } from "enzyme";

import NavList from "../";
import DropdownContainer from "../DropdownContainer";
import ItemContainer from "../ItemContainer";

/* eslint-disable react/prop-types */
const Item = ({ text }) => <div>{text}</div>;

const Dropdown = ({ nav }) => <li>{nav.map(n => <ul>{n.text}</ul>)}</li>;

describe("NavList", () => {
  const toggleDropdown = jest.fn();

  const data = {
    toggleDropdown,
    navItem: Item,
    dropdown: Dropdown,
    nav: [
      {
        to: "#",
        text: "foo",
      },
      {
        to: "#",
        text: "bar",
        open: true,
        subnav: [
          {
            to: "biz",
            text: "baz",
          },
        ],
      },
      {
        to: "#",
        text: "bang",
        subnav: [
          {
            to: "boom",
            text: "boom",
          },
        ],
      },
    ],
  };
  test("renders correctly", () => {
    const wrapper = shallow(<NavList {...data} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("item container", () => {
    const wrapper = shallow(<ItemContainer>foo</ItemContainer>);
    expect(wrapper).toMatchSnapshot();
  });

  test("dropdown container", () => {
    const wrapper = shallow(<DropdownContainer>foo</DropdownContainer>);
    expect(wrapper).toMatchSnapshot();
  });
});
