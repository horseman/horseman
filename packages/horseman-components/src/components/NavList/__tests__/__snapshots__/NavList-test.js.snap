// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavList dropdown container 1`] = `
.c0 {
  position: absolute;
  top: 100%;
  left: 0;
  z-index: 10;
}

<div
  className="c0"
>
  foo
</div>
`;

exports[`NavList item container 1`] = `
.c0 {
  position: relative;
  overflow: visible;
  margin-right: 1em;
}

.c0:last-child {
  margin-right: 0;
}

<li
  className="c0"
>
  foo
</li>
`;

exports[`NavList renders correctly 1`] = `
<styled.ul>
  <styled.li
    key="foo"
  >
    <navItem
      onClick={[Function]}
      text="foo"
      to="#"
    />
  </styled.li>
  <styled.li
    key="bar"
  >
    <navItem
      onClick={[Function]}
      subnav={
        Array [
          Object {
            "text": "baz",
            "to": "biz",
          },
        ]
      }
      text="bar"
      to="#"
    />
  </styled.li>
</styled.ul>
`;
